modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,343 @@
+package proto
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/errors"
+)
+
+// merkleRootsCacheHeight is the height of the subTrees in cachedSubTrees. A
+// height of 7 means that 128 sector roots are covered by a single cached
+// subTree.
+const merkleRootsCacheHeight = 7
+
+type (
+	// merkleRoots is a helper struct that makes it easier to add/insert/remove
+	// merkleRoots within a SafeContract.
+	merkleRoots struct {
+		// cachedSubTrees are cached trees that can be used to more efficiently
+		// compute the merkle root of a contract. The cached trees are not
+		// persisted and are computed after startup.
+		cachedSubTrees []*cachedSubTree
+		// uncachedRoots contains the sector roots that are not part of a
+		// cached subTree. The uncachedRoots slice should never get longer than
+		// 2^merkleRootsCacheHeight since that would simply result in a new
+		// cached subTree in cachedSubTrees.
+		uncachedRoots []crypto.Hash
+
+		// file is the file of the safe contract that contains the root.
+		file *os.File
```

Comment:
I don't like that both `merkleRoots` and `SafeContract` have a handle to the same file. If you're just looking at the `SafeContract` code, it's not obvious that calling a method on `sc.merkleRoots` modifies sc.f`. I'm not sure what the best way to resolve this is. Maybe it's a mistake to have a separate `merkleRoots` type, and it should just be merged with `SafeContract`.

Reply:
going to agree with this, shouldn't have 2 objects with atomic assumptions pointing to the same file on disk, they might break eachother.

We should have this all happen under the SafeContract, using the wal of the SafeContract.